src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs

```diff
@@ -0,0 +1,117 @@
+            RuleLevel.BuildWarning,
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotUseWaitAllWithSingleTaskDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(WhenAllRule, WaitAllRule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                if (compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask, out var taskType) &&
+                    compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask1, out var task1Type))
+                {
+                    compilationContext.RegisterOperationAction(operationContext =>
+                    {
+                        var invocation = (IInvocationOperation)operationContext.Operation;
+                        if (IsWhenOrWaitAllMethod(invocation.TargetMethod, taskType) &&
+                            IsSingleTaskArgument(invocation, taskType, task1Type, operationContext.CancellationToken))
+                        {
+                            switch (invocation.TargetMethod.Name)
+                            {
+                                case nameof(Task.WhenAll):
+                                    operationContext.ReportDiagnostic(invocation.CreateDiagnostic(WhenAllRule));
+                                    break;
+
+                                case nameof(Task.WaitAll):
+                                    operationContext.ReportDiagnostic(invocation.CreateDiagnostic(WaitAllRule));
+                                    break;
+
+                                default:
+                                    throw new InvalidOperationException($"Unexpected method name: {invocation.TargetMethod.Name}");
+                            }
+                        }
+                    }, OperationKind.Invocation);
+                }
+            });
+        }
+
+        private static bool IsWhenOrWaitAllMethod(IMethodSymbol targetMethod, INamedTypeSymbol taskType)
+        {
+            var nameMatches = targetMethod.Name is (nameof(Task.WhenAll)) or (nameof(Task.WaitAll));
+            var parameters = targetMethod.Parameters;
+
+            return nameMatches &&
+                targetMethod.IsStatic &&
+                SymbolEqualityComparer.Default.Equals(targetMethod.ContainingType, taskType) &&
+                parameters.Length == 1 &&
+                parameters[0].IsParams;
+        }
```

Comment:
Not critical, but I think a bit more performant implementation would be to compute and store the method symbols for `Task.WhenAll` and `Task.WaitAll` upfront in the compilation start action and then this method would just need to perform symbol equality checks.